lib/highline.rb

```diff
@@ -717,105 +485,175 @@ def output_cols
+    say(question.responses[error]) unless error.nil?
+    say(question.ask_on_error_msg)
+  end
 
-    template  = ERB.new(statement, nil, "%")
-    statement = template.result(binding)
+  #
+  # Gets one answer, as opposed to HighLine#gather
+  #
+  def ask_once(question)
+
+    # readline() needs to handle its own output, but readline only supports
+    # full line reading.  Therefore if question.echo is anything but true,
+    # the prompt will not be issued. And we have to account for that now.
+    # Also, JRuby-1.7's ConsoleReader.readLine() needs to be passed the prompt
+    # to handle line editing properly.
+    say(question) unless ((question.readline) and (question.echo == true and question.limit.nil?))
 
-    statement = wrap(statement) unless @wrap_at.nil?
-    statement = page_print(statement) unless @page_at.nil?
+    begin
+      question.get_response_or_default(self)
+      raise NotValidQuestionError unless question.valid_answer?
 
-    # 'statement' is encoded in US-ASCII when using ruby 1.9.3(-p551)
-    # 'indentation' is correctly encoded (same as default_external encoding)
-    statement = statement.force_encoding(Encoding.default_external)
+      question.convert
 
-    statement = statement.gsub(/\n(?!$)/,"\n#{indentation}") if @multi_indent
+      if question.confirm
+        # need to add a layer of scope (new_scope) to ask a question inside a
+        # question, without destroying instance data
 
-    statement
-  end
+        raise NoConfirmationQuestionError unless confirm(question)
+      end
 
-  #
-  # A helper method for sending the output stream and error and repeat
-  # of the question.
-  #
-  def explain_error( error )
-    say(@question.responses[error]) unless error.nil?
-    if @question.responses[:ask_on_error] == :question
-      say(@question)
-    elsif @question.responses[:ask_on_error]
-      say(@question.responses[:ask_on_error])
+    rescue NoConfirmationQuestionError
+      explain_error(nil, question)
+      retry
+
+    rescue NotInRangeQuestionError
+      explain_error(:not_in_range, question)
+      retry
+
+    rescue NotValidQuestionError
+      explain_error(:not_valid, question)
+      retry
+
+    rescue QuestionError
+      retry
+
+    rescue ArgumentError => error
+      case error.message
+      when /ambiguous/
+        # the assumption here is that OptionParser::Completion#complete
+        # (used for ambiguity resolution) throws exceptions containing
+        # the word 'ambiguous' whenever resolution fails
+        explain_error(:ambiguous_completion, question)
+        retry
+      when /invalid value for/
+        explain_error(:invalid_type, question)
+        retry
+      else
+        raise
+      end
+
+    rescue Question::NoAutoCompleteMatch
+      explain_error(:no_completion, question)
+      retry
     end
+    question.answer
   end
 
+  def confirm(question)
+    new_scope.agree(question.confirm_question(self))
+  end
+
+
+  public :ask_once
```

Comment:
What about remove this `public :ask_once` and move the method outside the private?